src/core/packages/elasticsearch/client-server-mocks/src/mocks.ts

```diff
@@ -122,67 +121,176 @@ const createMockedApi = <
+    Object.defineProperty(obj, key, {
       configurable: true,
-      enumerable: false,
-      get: () => jest.fn(),
-      set: undefined,
+      enumerable: true,
+      get() {
+        const fn = createMockedApi();
+        fn.mockImplementation(() => res ?? createSuccessTransportRequestPromise({}));
+        Object.defineProperty(obj, key, {
+          value: fn,
+          configurable: true,
+          enumerable: true,
+          writable: true,
+        });
+        return fn;
+      },
+      set(value) {
+        Object.defineProperty(obj, key, {
+          value,
+          configurable: true,
+          enumerable: true,
+          writable: true,
+        });
+      },
     });
   };
 
-  // `on`, `off`, and `once` are properties without a setter.
-  // We can't `client.diagnostic.on = jest.fn()` because the following error will be thrown:
-  // TypeError: Cannot set property on of #<Client> which has only a getter
-  mockGetter(client.diagnostic, 'on');
-  mockGetter(client.diagnostic, 'off');
-  mockGetter(client.diagnostic, 'once');
-  client.transport = {
-    request: jest.fn(),
+  const defineLazyObject = (obj: Record<string, any>, key: string, childShape: ShapeNode) => {
+    Object.defineProperty(obj, key, {
+      configurable: true,
+      enumerable: true,
+      get() {
+        const child = buildLazyMockFromShape(childShape, res);
+        Object.defineProperty(obj, key, {
+          value: child,
+          configurable: true,
+          enumerable: true,
+          writable: true,
+        });
+        return child;
+      },
+    });
   };
 
-  return client as DeeplyMockedApi<Client>;
+  for (const [key, node] of Object.entries(shape.props)) {
+    if (node.type === 'method') {
+      defineLazyMethod(target, key);
+    } else if (node.type === 'object') {
+      defineLazyObject(target, key, node);
+    }
+  }
+
+  // Special cases based on prior behavior
+  Object.defineProperty(target, 'diagnostic', {
+    configurable: true,
+    enumerable: false,
+    get() {
+      const d: any = {};
+      for (const k of ['on', 'off', 'once']) {
+        Object.defineProperty(d, k, {
+          configurable: true,
+          enumerable: false,
+          get: () => jest.fn(),
+        });
+      }
+      Object.defineProperty(target, 'diagnostic', {
+        value: d,
+        configurable: true,
+        enumerable: false,
+        writable: true,
+      });
+      return d;
+    },
+  });
+
+  Object.defineProperty(target, 'transport', {
+    configurable: true,
+    enumerable: true,
+    get() {
+      const t = { request: jest.fn() } as any;
+      Object.defineProperty(target, 'transport', {
+        value: t,
+        configurable: true,
+        enumerable: true,
+        writable: true,
+      });
+      return t;
+    },
+  });
+
+  return target;
+}
+
+const createInternalClientMock = (res?: Promise<unknown>): DeeplyMockedApi<Client> => {
```

Comment:
I wonder if it would be cleaner to default `res = createSuccessTransportRequestPromise({})` here, and make it mandatory in `buildLazyMockFromShape`.

I believe it would create a lot less Promises.

Reply:
Done in [1cc85e0](https://github.com/elastic/kibana/pull/236883/commits/1cc85e0d190b022d9c254afb466577071e0090d0)